File: src/cache/hybrid_cache.cpp
#include "meteor/cache/hybrid_cache.hpp"
#include "meteor/storage/optimized_ssd_storage.hpp"
#include "meteor/utils/memory_pool.hpp"
#include "meteor/utils/thread_pool.hpp"
#include <algorithm>
#include <chrono>
#include <cstring>
#include <iostream>

namespace meteor {
namespace cache {

HybridCache::HybridCache(const Config& config)
    : config_(config)
    , memory_cache_(config.max_memory_entries)
    , memory_usage_(0)
    , hit_count_(0)
    , miss_count_(0)
    , eviction_count_(0)
    , running_(false)
    , memory_pool_(std::make_unique<utils::MemoryPool>(config.buffer_pool_size))
    , thread_pool_(std::make_unique<utils::ThreadPool>(config.io_threads))
{
    // Initialize SSD storage if tiered_prefix is provided
    if (!config.tiered_prefix.empty()) {
        try {
            ssd_storage_ = std::make_unique<storage::OptimizedSSDStorage>(
                config.tiered_prefix,
                config.page_size,
                config.max_file_size
            );
            use_ssd_ = true;
        } catch (const std::exception& e) {
            std::cerr << "Failed to initialize SSD storage: " << e.what() << std::endl;
            use_ssd_ = false;
        }
    } else {
        use_ssd_ = false;
    }
}

HybridCache::~HybridCache() {
    stop();
}

void HybridCache::start() {
    if (running_.exchange(true)) {
        return; // Already running
    }

    // Start cleanup thread
    cleanup_thread_ = std::thread([this]() {
        while (running_) {
            std::this_thread::sleep_for(std::chrono::seconds(config_.cleanup_interval));
            cleanup_expired_entries();
        }
    });

    // Start metrics thread
    metrics_thread_ = std::thread([this]() {
        while (running_) {
            std::this_thread::sleep_for(std::chrono::seconds(config_.metrics_interval));
            update_metrics();
        }
    });
}

void HybridCache::stop() {
    if (!running_.exchange(false)) {
        return; // Already stopped
    }

    if (cleanup_thread_.joinable()) {
        cleanup_thread_.join();
    }
    if (metrics_thread_.joinable()) {
        metrics_thread_.join();
    }

    // Sync SSD storage
    if (ssd_storage_) {
        ssd_storage_->sync();
    }
}

std::optional<std::string> HybridCache::get(const std::string& key) {
    auto start_time = std::chrono::steady_clock::now();

    // Try memory cache first
    {
        std::shared_lock<std::shared_mutex> lock(memory_mutex_);
        auto it = memory_cache_.find(key);
        if (it != memory_cache_.end()) {
            // Check if expired
            if (std::chrono::steady_clock::now() < it->second.expiry_time) {
                // Update access time for LRU
                it->second.last_access = std::chrono::steady_clock::now();
                hit_count_++;
                update_get_latency(start_time);
                return it->second.value;
            } else {
                // Expired, remove from memory
                memory_usage_ -= it->second.value.size();
                memory_cache_.erase(it);
            }
        }
    }

    // Try SSD storage if available
    if (use_ssd_ && ssd_storage_) {
        try {
            auto result = ssd_storage_->read_async(key);
            if (result.has_value()) {
                // Update memory cache
                put_memory_cache(key, result.value(), config_.default_ttl);
                hit_count_++;
                update_get_latency(start_time);
                return result.value();
            }
        } catch (const std::exception& e) {
            std::cerr << "SSD read error: " << e.what() << std::endl;
        }
    }

    miss_count_++;
    update_get_latency(start_time);
    return std::nullopt;
}

bool HybridCache::put(const std::string& key, const std::string& value, 
                      std::chrono::seconds ttl) {
    auto start_time = std::chrono::steady_clock::now();

    // Use default TTL if not specified
    if (ttl.count() <= 0) {
        ttl = config_.default_ttl;
    }

    // Always try to put in memory cache first
    bool memory_success = put_memory_cache(key, value, ttl);

    // Write to SSD storage asynchronously if available
    if (use_ssd_ && ssd_storage_) {
        thread_pool_->enqueue([this, key, value, ttl]() {
            try {
                ssd_storage_->write_async(key, value, ttl);
            } catch (const std::exception& e) {
                std::cerr << "SSD write error: " << e.what() << std::endl;
            }
        });
    }

    update_put_latency(start_time);
    return memory_success || use_ssd_;
}

bool HybridCache::remove(const std::string& key) {
    auto start_time = std::chrono::steady_clock::now();

    bool removed = false;

    // Remove from memory cache
    {
        std::unique_lock<std::shared_mutex> lock(memory_mutex_);
        auto it = memory_cache_.find(key);
        if (it != memory_cache_.end()) {
            memory_usage_ -= it->second.value.size();
            memory_cache_.erase(it);
            removed = true;
        }
    }

    // Note: We don't remove from SSD storage as it's append-only
    // The entry will be ignored due to TTL or overwritten on next write

    update_remove_latency(start_time);
    return removed;
}

bool HybridCache::contains(const std::string& key) {
    // Check memory cache first
    {
        std::shared_lock<std::shared_mutex> lock(memory_mutex_);
        auto it = memory_cache_.find(key);
        if (it != memory_cache_.end()) {
            if (std::chrono::steady_clock::now() < it->second.expiry_time) {
                return true;
            }
        }
    }

    // Check SSD storage if available
    if (use_ssd_ && ssd_storage_) {
        try {
            auto result = ssd_storage_->read_async(key);
            return result.has_value();
        } catch (const std::exception& e) {
            std::cerr << "SSD contains error: " << e.what() << std::endl;
        }
    }

    return false;
}

void HybridCache::clear() {
    {
        std::unique_lock<std::shared_mutex> lock(memory_mutex_);
        memory_cache_.clear();
        memory_usage_ = 0;
    }

    // Note: We don't clear SSD storage as it would require recreating files
    // Entries will expire naturally based on TTL
}

CacheStats HybridCache::get_stats() const {
    CacheStats stats;
    stats.memory_hits = hit_count_.load();
    stats.memory_misses = miss_count_.load();
    stats.memory_entries = memory_cache_.size();
    stats.memory_usage = memory_usage_.load();
    stats.evictions = eviction_count_.load();
    stats.ssd_enabled = use_ssd_;
    
    if (ssd_storage_) {
        auto ssd_stats = ssd_storage_->get_stats();
        stats.ssd_reads = ssd_stats.reads.load();
        stats.ssd_writes = ssd_stats.writes.load();
        stats.ssd_errors = ssd_stats.errors.load();
    }
    
    return stats;
}

bool HybridCache::put_memory_cache(const std::string& key, const std::string& value, 
                                   std::chrono::seconds ttl) {
    std::unique_lock<std::shared_mutex> lock(memory_mutex_);

    // Check if we need to evict entries
    while (memory_cache_.size() >= config_.max_memory_entries || 
           memory_usage_ + value.size() > config_.max_memory_bytes) {
        if (!evict_lru_entry()) {
            break; // No more entries to evict
        }
    }

    // Check if we still have space
    if (memory_cache_.size() >= config_.max_memory_entries || 
        memory_usage_ + value.size() > config_.max_memory_bytes) {
        return false; // No space available
    }

    // Remove existing entry if present
    auto it = memory_cache_.find(key);
    if (it != memory_cache_.end()) {
        memory_usage_ -= it->second.value.size();
    }

    // Add new entry
    CacheEntry entry;
    entry.value = value;
    entry.last_access = std::chrono::steady_clock::now();
    entry.expiry_time = entry.last_access + ttl;

    memory_cache_[key] = std::move(entry);
    memory_usage_ += value.size();

    return true;
}

bool HybridCache::evict_lru_entry() {
    if (memory_cache_.empty()) {
        return false;
    }

    // Find LRU entry
    auto lru_it = memory_cache_.begin();
    auto oldest_time = lru_it->second.last_access;

    for (auto it = memory_cache_.begin(); it != memory_cache_.end(); ++it) {
        if (it->second.last_access < oldest_time) {
            oldest_time = it->second.last_access;
            lru_it = it;
        }
    }

    // Remove LRU entry
    memory_usage_ -= lru_it->second.value.size();
    memory_cache_.erase(lru_it);
    eviction_count_++;

    return true;
}

void HybridCache::cleanup_expired_entries() {
    std::unique_lock<std::shared_mutex> lock(memory_mutex_);
    
    auto now = std::chrono::steady_clock::now();
    auto it = memory_cache_.begin();
    
    while (it != memory_cache_.end()) {
        if (now >= it->second.expiry_time) {
            memory_usage_ -= it->second.value.size();
            it = memory_cache_.erase(it);
        } else {
            ++it;
        }
    }
}

void HybridCache::update_metrics() {
    // Update any additional metrics here
    // This could include histogram updates, etc.
}

void HybridCache::update_get_latency(std::chrono::steady_clock::time_point start) {
    auto duration = std::chrono::steady_clock::now() - start;
    // Update latency histogram if available
}

void HybridCache::update_put_latency(std::chrono::steady_clock::time_point start) {
    auto duration = std::chrono::steady_clock::now() - start;
    // Update latency histogram if available
}

void HybridCache::update_remove_latency(std::chrono::steady_clock::time_point start) {
    auto duration = std::chrono::steady_clock::now() - start;
    // Update latency histogram if available
}

} // namespace cache
} // namespace meteor 

File: pkg/cache/hybrid_cache.go
package cache

import (
	"context"
	"fmt"
	"sync"
	"sync/atomic"
	"time"

	"github.com/kumar-sumit/meteor/pkg/storage"
)

// HybridCache implements a high-performance cache with optional SSD overflow
type HybridCache struct {
	// In-memory cache (L1)
	memoryCache map[string]*HybridCacheEntry
	memoryMu    sync.RWMutex

	// SSD cache (L2) - optional based on configuration
	ssdStorage *storage.OptimizedSSDStorage
	ssdEnabled bool

	// Configuration
	maxMemoryEntries   int
	maxMemoryBytes     int64
	currentMemoryBytes int64

	// LRU tracking for memory cache
	lruHead *HybridCacheEntry
	lruTail *HybridCacheEntry

	// Metrics
	memoryHits   int64
	memoryMisses int64
	ssdHits      int64
	ssdMisses    int64
	evictions    int64

	// Lifecycle
	ctx    context.Context
	cancel context.CancelFunc
	wg     sync.WaitGroup

	// Background processes
	evictionTicker *time.Ticker
	metricsTicker  *time.Ticker
}

// HybridCacheEntry represents a cache entry with LRU tracking
type HybridCacheEntry struct {
	key       string
	value     []byte
	timestamp time.Time
	ttl       time.Duration
	accessed  time.Time
	inMemory  bool
	size      int64

	// LRU linked list pointers
	prev *HybridCacheEntry
	next *HybridCacheEntry
}

// HybridCacheConfig configures the hybrid cache
type HybridCacheConfig struct {
	// Memory cache configuration
	MaxMemoryEntries int
	MaxMemoryBytes   int64

	// Tiered storage configuration
	TieredPrefix string // If empty, pure in-memory cache; if set, enables SSD tier

	// SSD cache configuration (used only when TieredPrefix is set)
	SSDPageSize    int
	SSDMaxFileSize int64

	// Performance tuning
	EvictionInterval time.Duration
	MetricsInterval  time.Duration

	// Default TTL
	DefaultTTL time.Duration
}

// DefaultHybridCacheConfig returns a default configuration for pure in-memory cache
func DefaultHybridCacheConfig() *HybridCacheConfig {
	return &HybridCacheConfig{
		MaxMemoryEntries: 100000,            // 100K entries in memory
		MaxMemoryBytes:   256 * 1024 * 1024, // 256MB memory limit
		TieredPrefix:     "",                // Pure in-memory by default
		SSDPageSize:      4096,
		SSDMaxFileSize:   2 * 1024 * 1024 * 1024, // 2GB per file
		EvictionInterval: time.Second * 30,
		MetricsInterval:  time.Second * 10,
		DefaultTTL:       time.Hour,
	}
}

// NewHybridCache creates a new hybrid cache instance
func NewHybridCache(config *HybridCacheConfig) (*HybridCache, error) {
	if config == nil {
		config = DefaultHybridCacheConfig()
	}

	ctx, cancel := context.WithCancel(context.Background())

	cache := &HybridCache{
		memoryCache:      make(map[string]*HybridCacheEntry),
		ssdEnabled:       config.TieredPrefix != "",
		maxMemoryEntries: config.MaxMemoryEntries,
		maxMemoryBytes:   config.MaxMemoryBytes,
		ctx:              ctx,
		cancel:           cancel,
		evictionTicker:   time.NewTicker(config.EvictionInterval),
		metricsTicker:    time.NewTicker(config.MetricsInterval),
	}

	// Create SSD storage backend only if TieredPrefix is configured
	if cache.ssdEnabled {
		ssdStorage, err := storage.NewOptimizedSSDStorage(
			config.TieredPrefix,
			config.SSDPageSize,
			config.SSDMaxFileSize,
		)
		if err != nil {
			cancel()
			return nil, fmt.Errorf("failed to create SSD storage: %w", err)
		}
		cache.ssdStorage = ssdStorage
	}

	// Initialize LRU list
	cache.lruHead = &HybridCacheEntry{}
	cache.lruTail = &HybridCacheEntry{}
	cache.lruHead.next = cache.lruTail
	cache.lruTail.prev = cache.lruHead

	// Start background processes
	backgroundWorkers := 2
	if cache.ssdEnabled {
		backgroundWorkers = 3
	}

	cache.wg.Add(backgroundWorkers)
	go cache.evictionWorker()
	go cache.metricsWorker()
	if cache.ssdEnabled {
		go cache.ssdWorker()
	}

	return cache, nil
}

// Get retrieves a value from the cache
func (c *HybridCache) Get(ctx context.Context, key string) ([]byte, error) {
	// Try memory cache first (L1)
	c.memoryMu.RLock()
	if entry, exists := c.memoryCache[key]; exists {
		if !c.isExpired(entry) {
			// Move to front of LRU
			c.moveToFront(entry)
			entry.accessed = time.Now()
			atomic.AddInt64(&c.memoryHits, 1)
			value := make([]byte, len(entry.value))
			copy(value, entry.value)
			c.memoryMu.RUnlock()
			return value, nil
		} else {
			// Entry expired, remove it
			c.memoryMu.RUnlock()
			c.memoryMu.Lock()
			c.removeFromMemory(key)
			c.memoryMu.Unlock()
			atomic.AddInt64(&c.memoryMisses, 1)
		}
	} else {
		c.memoryMu.RUnlock()
		atomic.AddInt64(&c.memoryMisses, 1)
	}

	// Try SSD cache (L2) only if enabled
	if c.ssdEnabled {
		value, err := c.ssdStorage.ReadAsync(key)
		if err != nil {
			atomic.AddInt64(&c.ssdMisses, 1)
			return nil, err
		}

		if value != nil {
			atomic.AddInt64(&c.ssdHits, 1)
			// Promote to memory cache
			c.promoteToMemory(key, value, time.Hour) // Default TTL
			return value, nil
		}

		atomic.AddInt64(&c.ssdMisses, 1)
	}

	return nil, nil
}

// Put stores a value in the cache
func (c *HybridCache) Put(ctx context.Context, key string, value []byte, ttl time.Duration) error {
	if ttl <= 0 {
		ttl = time.Hour // Default TTL
	}

	// Try to store in memory first (this is the fast path)
	if c.canFitInMemory(key, value) {
		// Fast path: store directly in memory without SSD write
		c.promoteToMemory(key, value, ttl)
		return nil
	}

	// Slow path: memory is full, need to use SSD in hybrid mode
	if c.ssdEnabled {
		// Write to SSD for persistence since memory is full
		if err := c.ssdStorage.WriteAsync(key, value); err != nil {
			return fmt.Errorf("failed to write to SSD: %w", err)
		}
		// Still try to store in memory (may cause eviction)
		c.promoteToMemory(key, value, ttl)
	} else {
		// Pure in-memory mode: just store in memory (may cause eviction)
		c.promoteToMemory(key, value, ttl)
	}

	return nil
}

// canFitInMemory checks if a key-value pair can fit in memory without eviction
func (c *HybridCache) canFitInMemory(key string, value []byte) bool {
	entrySize := int64(len(key) + len(value) + 128) // Approximate overhead

	c.memoryMu.RLock()
	currentEntries := len(c.memoryCache)
	c.memoryMu.RUnlock()

	currentMemoryBytes := atomic.LoadInt64(&c.currentMemoryBytes)

	// Check if we have space without eviction
	return currentEntries < c.maxMemoryEntries &&
		currentMemoryBytes+entrySize <= c.maxMemoryBytes
}

// Remove deletes a value from the cache
func (c *HybridCache) Remove(ctx context.Context, key string) error {
	// Remove from memory cache
	c.memoryMu.Lock()
	c.removeFromMemory(key)
	c.memoryMu.Unlock()

	// Note: We don't remove from SSD as it's append-only
	// The entry will be ignored due to TTL or overwritten

	return nil
}

// Contains checks if a key exists in the cache
func (c *HybridCache) Contains(ctx context.Context, key string) (bool, error) {
	// Check memory cache first
	c.memoryMu.RLock()
	if entry, exists := c.memoryCache[key]; exists {
		if !c.isExpired(entry) {
			c.memoryMu.RUnlock()
			return true, nil
		}
	}
	c.memoryMu.RUnlock()

	// Check SSD cache
	if c.ssdEnabled {
		value, err := c.ssdStorage.ReadAsync(key)
		if err != nil {
			return false, err
		}

		return value != nil, nil
	}

	return false, nil
}

// Clear removes all entries from the cache
func (c *HybridCache) Clear(ctx context.Context) error {
	c.memoryMu.Lock()
	defer c.memoryMu.Unlock()

	// Clear memory cache
	c.memoryCache = make(map[string]*HybridCacheEntry)
	c.lruHead.next = c.lruTail
	c.lruTail.prev = c.lruHead
	atomic.StoreInt64(&c.currentMemoryBytes, 0)

	// Note: We don't clear SSD storage as it would require file recreation

	return nil
}

// promoteToMemory promotes a key-value pair to memory cache
func (c *HybridCache) promoteToMemory(key string, value []byte, ttl time.Duration) {
	c.memoryMu.Lock()
	defer c.memoryMu.Unlock()

	entrySize := int64(len(key) + len(value) + 128) // Approximate overhead

	// Check if we need to evict entries
	for len(c.memoryCache) >= c.maxMemoryEntries ||
		atomic.LoadInt64(&c.currentMemoryBytes)+entrySize > c.maxMemoryBytes {

		if !c.evictLRUToSSD() {
			break // No more entries to evict
		}
	}

	// Create new entry
	entry := &HybridCacheEntry{
		key:       key,
		value:     make([]byte, len(value)),
		timestamp: time.Now(),
		ttl:       ttl,
		accessed:  time.Now(),
		inMemory:  true,
		size:      entrySize,
	}
	copy(entry.value, value)

	// Add to memory cache
	c.memoryCache[key] = entry
	c.addToFront(entry)
	atomic.AddInt64(&c.currentMemoryBytes, entrySize)
}

// evictLRUToSSD evicts the least recently used entry, optionally writing to SSD
func (c *HybridCache) evictLRUToSSD() bool {
	if c.lruTail.prev == c.lruHead {
		return false // No entries to evict
	}

	entry := c.lruTail.prev

	// Write to SSD before evicting from memory (hybrid mode only)
	if c.ssdEnabled {
		// Write evicted entry to SSD asynchronously
		go func(key string, value []byte) {
			if err := c.ssdStorage.WriteAsync(key, value); err != nil {
				// Log error but don't block eviction
				// In production, you might want to use a proper logger
				fmt.Printf("Failed to write evicted entry to SSD: %v\n", err)
			}
		}(entry.key, entry.value)
	}

	c.removeFromMemory(entry.key)
	atomic.AddInt64(&c.evictions, 1)
	return true
}

// evictLRU evicts the least recently used entry (backward compatibility)
func (c *HybridCache) evictLRU() bool {
	return c.evictLRUToSSD()
}

// removeFromMemory removes an entry from memory cache
func (c *HybridCache) removeFromMemory(key string) {
	if entry, exists := c.memoryCache[key]; exists {
		delete(c.memoryCache, key)
		c.removeFromLRU(entry)
		atomic.AddInt64(&c.currentMemoryBytes, -entry.size)
	}
}

// LRU list management
func (c *HybridCache) addToFront(entry *HybridCacheEntry) {
	entry.next = c.lruHead.next
	entry.prev = c.lruHead
	c.lruHead.next.prev = entry
	c.lruHead.next = entry
}

func (c *HybridCache) removeFromLRU(entry *HybridCacheEntry) {
	entry.prev.next = entry.next
	entry.next.prev = entry.prev
}

func (c *HybridCache) moveToFront(entry *HybridCacheEntry) {
	c.removeFromLRU(entry)
	c.addToFront(entry)
}

// isExpired checks if an entry is expired
func (c *HybridCache) isExpired(entry *HybridCacheEntry) bool {
	return time.Since(entry.timestamp) > entry.ttl
}

// Background workers
func (c *HybridCache) evictionWorker() {
	defer c.wg.Done()

	for {
		select {
		case <-c.ctx.Done():
			return
		case <-c.evictionTicker.C:
			c.performEviction()
		}
	}
}

func (c *HybridCache) metricsWorker() {
	defer c.wg.Done()

	for {
		select {
		case <-c.ctx.Done():
			return
		case <-c.metricsTicker.C:
			c.logMetrics()
		}
	}
}

func (c *HybridCache) ssdWorker() {
	defer c.wg.Done()

	// This worker can handle background SSD maintenance tasks
	ticker := time.NewTicker(time.Minute)
	defer ticker.Stop()

	for {
		select {
		case <-c.ctx.Done():
			return
		case <-ticker.C:
			// Perform SSD maintenance (e.g., compaction, cleanup)
			c.performSSDMaintenance()
		}
	}
}

// performEviction removes expired entries from memory cache
func (c *HybridCache) performEviction() {
	c.memoryMu.Lock()
	defer c.memoryMu.Unlock()

	now := time.Now()
	expiredKeys := make([]string, 0)

	for key, entry := range c.memoryCache {
		if now.Sub(entry.timestamp) > entry.ttl {
			expiredKeys = append(expiredKeys, key)
		}
	}

	for _, key := range expiredKeys {
		c.removeFromMemory(key)
	}
}

// performSSDMaintenance performs background SSD maintenance
func (c *HybridCache) performSSDMaintenance() {
	// Placeholder for SSD maintenance tasks
	// Could include compaction, cleanup, etc.
}

// logMetrics logs performance metrics
func (c *HybridCache) logMetrics() {
	memHits := atomic.LoadInt64(&c.memoryHits)
	memMisses := atomic.LoadInt64(&c.memoryMisses)
	ssdHits := atomic.LoadInt64(&c.ssdHits)
	ssdMisses := atomic.LoadInt64(&c.ssdMisses)
	evictions := atomic.LoadInt64(&c.evictions)

	memoryBytes := atomic.LoadInt64(&c.currentMemoryBytes)

	c.memoryMu.RLock()
	memoryEntries := len(c.memoryCache)
	c.memoryMu.RUnlock()

	fmt.Printf("HybridCache Metrics: MemHits=%d, MemMisses=%d, SSDHits=%d, SSDMisses=%d, Evictions=%d, MemEntries=%d, MemBytes=%d\n",
		memHits, memMisses, ssdHits, ssdMisses, evictions, memoryEntries, memoryBytes)
}

// GetStats returns cache statistics
func (c *HybridCache) GetStats() map[string]interface{} {
	c.memoryMu.RLock()
	memoryEntries := len(c.memoryCache)
	c.memoryMu.RUnlock()

	return map[string]interface{}{
		"memory_hits":      atomic.LoadInt64(&c.memoryHits),
		"memory_misses":    atomic.LoadInt64(&c.memoryMisses),
		"ssd_hits":         atomic.LoadInt64(&c.ssdHits),
		"ssd_misses":       atomic.LoadInt64(&c.ssdMisses),
		"evictions":        atomic.LoadInt64(&c.evictions),
		"memory_entries":   memoryEntries,
		"memory_bytes":     atomic.LoadInt64(&c.currentMemoryBytes),
		"max_memory_bytes": c.maxMemoryBytes,
		"memory_hit_rate":  float64(atomic.LoadInt64(&c.memoryHits)) / float64(atomic.LoadInt64(&c.memoryHits)+atomic.LoadInt64(&c.memoryMisses)+1),
		"ssd_hit_rate":     float64(atomic.LoadInt64(&c.ssdHits)) / float64(atomic.LoadInt64(&c.ssdHits)+atomic.LoadInt64(&c.ssdMisses)+1),
	}
}

// Close shuts down the cache and releases resources
func (c *HybridCache) Close() error {
	c.cancel()
	c.wg.Wait()

	c.evictionTicker.Stop()
	c.metricsTicker.Stop()

	if c.ssdEnabled {
		if err := c.ssdStorage.Close(); err != nil {
			return fmt.Errorf("failed to close SSD storage: %w", err)
		}
	}

	return nil
}

// Additional methods for batch operations
func (c *HybridCache) BatchGet(ctx context.Context, keys []string) (map[string][]byte, error) {
	results := make(map[string][]byte)

	for _, key := range keys {
		if value, err := c.Get(ctx, key); err == nil && value != nil {
			results[key] = value
		}
	}

	return results, nil
}

func (c *HybridCache) BatchPut(ctx context.Context, entries map[string][]byte, ttl time.Duration) error {
	for key, value := range entries {
		if err := c.Put(ctx, key, value, ttl); err != nil {
			return err
		}
	}
	return nil
}

func (c *HybridCache) BatchRemove(ctx context.Context, keys []string) error {
	for _, key := range keys {
		if err := c.Remove(ctx, key); err != nil {
			return err
		}
	}
	return nil
}
